logging/logger.go

```diff
@@ -0,0 +1,134 @@
+package logging
+
+import (
+	"context"
+)
+
+var (
+	// DefaultLogger is the default Logger for all SDK logging, if one hasn't been provided in the context.
+	DefaultLogger Logger = &NoOpLogger{}
+
+	contextKey = loggerContextKey{}
+)
+
+type loggerContextKey struct{}
+
+// FromContext returns the Logger set in the context with Context(), or the DefaultLogger if no Logger is set in the context.
+// If DefaultLogger is nil, it returns a *NoOpLogger so that the return is always valid to call methods on without nil-checking.
+func FromContext(ctx context.Context) Logger {
```

Comment:
Seeing as this is going to be the usual pattern of using the logger I suggest adding the context to the logger here – this will make it so that calls like `logging.FromContext(ctx).Info(...)` will always carry the context down to e.g. `traceIDHandler` as opposed to requiring the user to call `logging.FromContext(ctx).InfoContext(...)`, especially since a user might not necessarily be aware or care about the latter invocation.

This means in practice that we'd need to add a new method to the logger interface, e.g. `WithContext` or similar. Interestingly enough, `slog` used to have such method at some point but it seems like they got rid of it upon release.

Reply:
Do we think it may be useful to drop the `<X>Context` methods from `Logger` and instead just have `WithContext(context.Context) Logger` which would ultimately achieve the same effect? Or is there value in having both?

Reply:
> Do we think it may be useful to drop the `<X>Context` methods from `Logger` and instead just have `WithContext(context.Context) Logger` which would ultimately achieve the same effect? Or is there value in having both?

I am in favour of dropping the extra methods and enforcing the `WithContext(...).Info(...)` approach.